example/mailto/mailto.cpp

```diff
@@ -0,0 +1,773 @@
+        return true;
+    }
+
+    static
+    bool
+    is_address(urls::string_view s) noexcept
+    {
+        // addr-spec = local-part "@" domain
+        std::size_t p = s.find('@');
+        if (p == urls::string_view::npos)
+            return false;
+        return is_local_part(s.substr(0, p)) &&
+               is_domain(s.substr(p + 1));
+    }
+
+    static
+    bool
+    is_local_part(urls::string_view s) noexcept
+    {
+        // local-part = dot-atom-text / quoted-string
+        return is_dot_atom_text(s) ||
+               is_quoted_string(s);
+    }
+
+    static
+    bool
+    is_dot_atom_text(urls::string_view s) noexcept
+    {
+        // dot-atom-text = 1*atext *("." 1*atext)
+        if (s.empty())
+            return false;
+        bool prev_dot = true;
+        for (std::size_t i = 0; i < s.size(); ++i)
+        {
+            if (prev_dot)
+            {
+                if (is_atext(s[i]))
+                    prev_dot = false;
+                else
+                    return false;
+            }
+            else if (!is_atext(s[i]))
+            {
+                if (s[i] == '.')
+                    prev_dot = true;
+                else
+                    return false;
+            }
+        }
+        return true;
+    }
+
+    static
+    bool
+    is_atext(char c) noexcept
+    {
+        // atext = ALPHA / DIGIT / "!" / "#" / "$" / "%" / "&" / "'" /
+        // "*" / "+" / "-" / "/" / "=" / "?" / "^" / "_" / "`" / "{" /
+        // "|" / "}" / "~"
+        return urls::grammar::alnum_chars(c) ||
+               c == '!'  ||
+               (c >= '#' && c <= '\'') ||
+               c == '*'  || c == '+'  ||
+               c == '-'  || c == '/'  ||
+               c == '='  || c == '?'  ||
+               (c >= '^' && c <= '`') ||
+               (c >= '{' && c <= '~');
+    }
+
+    static
+    bool
+    is_quoted_string(urls::string_view s) noexcept
+    {
+        // quoted-string = [CFWS] DQUOTE *([FWS] qcontent) [FWS] DQUOTE [CFWS]
+        auto it = s.data();
+        auto end = s.data() + s.size();
+        if (!parse_cfws(it, end))
+            return false;
+        if (it == end || *it++ != '"')
+            return false;
+        auto it0 = it;
+
+        // *([FWS] qcontent)
+        while (parse_fws_qcontent(it, end))
+        {
+            it0 = it;
+        }
+        it = it0;
+
+        // [FWS] DQUOTE [CFWS]
+        it0 = it;
+        if (!parse_fws(it, end))
+        {
+            it = it0;
+        }
+        if (it == end || *it++ != '"')
+            return false;
+        it0 = it;
+        if (!parse_cfws(it, end))
+        {
+            it = it0;
+        }
+        return true;
+    }
+
+    static
+    bool parse_fws_qcontent(char const*& it, char const* end) noexcept
+    {
+        // [FWS] qcontent
+        // [FWS]
+        auto it0 = it;
+        if (!parse_fws(it, end))
+        {
+            it = it0;
+        }
+        // qcontent = qtext / quoted-pair
```

Comment:
HTTP use quoted-pair as well... in rfc7230: https://www.rfc-editor.org/rfc/rfc7230#section-3.2.6

Reply:
nice

Reply:
:laughing: